### ISSUE TITLE
Make HotSpot builds for Linux/ARM32 production ready

### ISSUE THREAD

**Issue by sxa** [AdoptOpenJDK/openjdk-build#406]
I might split this into multiple issues since I suspect I'm seeing various underlying issues, but I'll start with this one to track what I'm trying to do ... List of current issues:
- [x] Builds are taking a long time (4-5 hours for JDK8/9, 13-14 hours for JDK10
- [x] ccache is enabled but doesn't appear to be taking effect in all cases (maybe missing C++) unless `/usr/lib/ccache` is put at the start of the path
- [x] Build scripts are currently defaulting to producing a zero VM so the VM is slow (Explains the first bullet point here) (Ref: https://github.com/AdoptOpenJDK/openjdk-build/pull/405)
- [ ] JDK8 builds appear to end up trying to build with `-march=i586` when you enable the JIT VM

**Comment by sxa555** [comment]
Now have a JIT-enabled openjdk9+hotspot installed in `/home/jenkins/jdk-9+181` on both armv7l machines and updated the `JDK9_BOOT_DIR` definitions in the jenkins configuration for the machines https://ci.adoptopenjdk.net/computer/build-scaleway-armv7-ubuntu16-04-2 and https://ci.adoptopenjdk.net/computer/build-scaleway-armv7-ubuntu16-04-1/ to point to it, so hopefully the next [openjdk10 build](https://ci.adoptopenjdk.net/job/openjdk10_build_arm32_linux/10/console) will use it

**Comment by sxa555** [comment]
Adjusted `JDK8_BOOT_DIR` to point to the default system one - /usr/lib/jvm/java-8-openjdk-armhf - instead of `/home/boot_jdk8` - which should speed up the [openjdk9 builds](https://ci.adoptopenjdk.net/job/openjdk9_build_arm32_linux/).

**Comment by sxa555** [comment]
Note that the system openjdk7 at `/usr/lib/jvm//usr/lib/jvm/java-1.7.0-openjdk-armhf` is a "zero" VM and therefore the 8 builds will still be slow unless we bootstrap 8 with 8 which I believe is possible ... Trying at https://ci.adoptopenjdk.net/job/openjdk8_build_arm32_linux/40 (Buiding zero VM)

**Comment by sxa555** [comment]
Bumped ccache size up from 1Gb to 2Gb on build-scaleway-ubuntu1604-armv7-1 (build-scaleway-ubuntu1604-armv7-2 in jenkins) to see if that helps things run faster going forward

**Comment by sxa555** [comment]
gzipping the tarball at the end takes >10 minutes which is quite a bit of extra time. Perhaps we should install and use `pigz` to use all four CPU cores.

**Comment by sxa555** [comment]
JDK8 build times at https://ci.adoptopenjdk.net/job/openjdk8_build_arm32_linux/buildTimeTrend down from 4 hours to about 1 hour after using a JITted JDK8 as the boot JDK. Further enhancements possible:
36 - 4h7 - zero VM openjdk7 boot JDK
38-39 - Attempt to build with `JVM_VARIANT=client` which failed trying to compile with `-march=i586`
40 - 50m - Using system JDK8 as boot JDK
41 - Attempt to build with `JVM_VARIANT=server`
42 - openjdk from build 40 used as the boot JDK
43+ - ccache size increased to 2Gb

**Comment by sxa555** [comment]
Both machines have now got adoptopenjdk8/9 JDKs in `/home/jenkins/bootjdks` and the `JDK?_BOOT_DIR` variables set to point to them.

**Comment by sxa555** [comment]
openjdk10 doesn't like me adding the ccache directory to the path - rude! :-)
```
checking for gcc... /usr/lib/ccache/gcc
checking resolved symbolic links for CC... /usr/bin/ccache
configure: Please use --enable-ccache instead of providing a wrapped compiler.
configure: error: /usr/lib/ccache/gcc is a symbolic link to ccache. This is not supported.
configure exiting with result code 1
```

**Comment by sxa555** [comment]
pigz changes integrated: https://github.com/AdoptOpenJDK/openjdk-build/pull/408
Added to openjdk10 pipelines (Although without testing as there are no current jobs/separate machines for testing as on AIX)
Build times are broadly comparable to what we're seeing on AIX so arm32 should not slow down any activities unless there are major changes which cause the ccache cache to be invalidated (may do one run of both 8 and 10 after clearing the ccache to get a baseline.
Only outstanding issue is why `-march=i586` shows up when building openjdk8 with the client or server VMs when building natively https://github.com/AdoptOpenJDK/openjdk-build/pull/405

**Comment by sxa555** [comment]
Default PIPELINE_TIMESTAMP wasn't being set on the AO9/HS builds - now fixed

**Comment by jerboaa** [comment]
>  checking for gcc... /usr/lib/ccache/gcc
> checking resolved symbolic links for CC... /usr/bin/ccache
> configure: Please use --enable-ccache instead of providing a wrapped compiler.
> configure: error: /usr/lib/ccache/gcc is a symbolic link to ccache. This is not supported.
> configure exiting with result code 1

Try using configure:

   configure [...] --enable-ccache

**Comment by sxa555** [comment]
Yep already set that back. It works for openjdk8 though

**Comment by sxa555** [comment]
@jerboaa Any idea why I can't seem to compile with `JVM_VARIANT=client` natively on arm32 without it trying to feed in `-march=i586` to the compiler? It seems to be acting as though it's cross-compiling. Fine on openjdk9 and 10 but not 8 at the moment

**Comment by jerboaa** [comment]
@sxa555 I'm commenting on #405 for the build issue.

**Comment by sxa555** [comment]
An update to this - https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk11u/job/jdk11u-linux-arm-hotspot/ is now running cleanly

**Comment by karianna** [comment]
So this can be closed now I think?

**Comment by sxa555** [comment]
I left it open on the basis that I don't believe we don't have regular JDK8 ones running at present

**Comment by niecore** [comment]
Is there a technical problem in building linux-aarch32hf (armv7/8/6) JDK8/HotSpot vm's?

**Comment by mpe85** [comment]
Will there be an JDK11 arm32 linux build officially provided via the website (https://adoptopenjdk.net/releases.html?variant=openjdk11&jvmVariant=hotspot), or will there only be snapshot builds that could be downloaded from the jenkins CI?

**Comment by karianna** [comment]
> Will there be an JDK11 arm32 linux build officially provided via the website (https://adoptopenjdk.net/releases.html?variant=openjdk11&jvmVariant=hotspot), or will there only be snapshot builds that could be downloaded from the jenkins CI?

The API and website should be showing nightly builds for that platform

**Comment by mpe85** [comment]
The API shows this build: OpenJDK11-jdk_arm_linux_hotspot_11_28.tar.gz
The website apparently doesn't show any arm32 builds (neither releases nor nightly builds).

When will the nightly builds of this platform become a release build?

**Comment by karianna** [comment]
> The API shows this build: OpenJDK11-jdk_arm_linux_hotspot_11_28.tar.gz
> The website apparently doesn't show any arm32 builds (neither releases nor nightly builds).

Hmm, odd that may be an API or website bug or the last nightly build for this was a long time ago.  Can you pop along to our slack and raise it in the API channel?
 
> When will the nightly builds of this platform become a release build?

With the upstream release cycles.  So 14th of April for 11.0.3

**Comment by sxa555** [comment]
Having machine stability issues which need investigating, along with getting the new machine set up and running:

FYI https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk11u/job/jdk11u-linux-arm-hotspot/86/consoleFull ran ok on build-scaleway-armv7-ubuntu16-04-2, https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk12u/job/jdk12u-linux-arm-hotspot/25/consoleFull ran on build-packet-ubuntu1604-armv7l-1, but all those machines are currently offline

My raspberry pi has crashed when running builds so I'm re-running to investigate now (since it's easy to restart if it dies!)

**Comment by sxa555** [comment]
Despite being set in jenkins with `NUM_PROCESSORS=4` my [JDK12 build](https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk12u/job/jdk12u-linux-arm-hotspot/36/console) is causing a load of nearly 40 on the machine (and jenkins has now disowned it) and there appears to be about 12-16 gcc processes running in parallel along with other stuff. It's 1.5Gb into swap space too ... Something's definitely not working as it should. I'll drop `NUM_PROCESSORS=1` for the next run

```
Creating support/modules_libs/java.base/server/libjvm.so from 844 file(s)
Creating hotspot/variant-server/libjvm/gtest/libjvm.so from 104 file(s)
Creating hotspot/variant-server/libjvm/gtest/gtestLauncher from 1 file(s)
Cannot contact build-sxa-armv7-raspbian: java.lang.InterruptedException
```

**Comment by M-Davies** [comment]
@sxa With https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk8u/job/jdk8u-linux-arm-hotspot/ and https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk14u/job/jdk14u-linux-arm-hotspot/ running smoothly, can this now be closed?

**Comment by sxa** [comment]
We are not building JDK8 properly (with the JIT) so I'd argue no

**Comment by sxa** [comment]
arm32 JDK8u with JIT is now built and shipping from https://github.com/AdoptOpenJDK/openjdk-aarch32-jdk8u, therefore closing.
